cpp/benchmarks/join/generate_input_tables.cuh

```diff
@@ -0,0 +1,275 @@
+ *                                  present in the build table.
+ * @param[in] rand_max              maximum random number to generate. I.e. random numbers are
+ *                                  integers from [0,rand_max].
+ * @param[in] uniq_build_tbl_keys   if each key in the build table should appear exactly once.
+ */
+template<typename key_type, typename size_type>
+void generate_input_tables(
+                           key_type* const build_tbl,
+                           const size_type build_tbl_size,
+                           key_type* const probe_tbl,
+                           const size_type probe_tbl_size,
+                           const double selectivity,
+                           const key_type rand_max,
+                           const bool uniq_build_tbl_keys)
+{
+    // With large values of rand_max the a lot of temporary storage is needed for the lottery. At the expense of not
+    // being that accurate with applying the selectivity an especially more memory efficient implementations would be
+    // to partition the random numbers into two intervals and then let one table choose random numbers from only one
+    // interval and the other only select with selectivity propability from the same interval and from the other in the
+    // other cases.
+
+    static_assert(std::is_signed<key_type>::value, "key_type needs to be signed for lottery to work");
+
+    const int block_size = 128;
+
+    // Maximize exposed parallelism while minimizing storage for curand state
+    int num_blocks_init_build_tbl {-1};
+    CUDA_TRY(cudaOccupancyMaxActiveBlocksPerMultiprocessor(
+        &num_blocks_init_build_tbl, init_build_tbl<key_type, size_type>, block_size, 0
+    ));
+
+    int num_blocks_init_probe_tbl {-1};
+    CUDA_TRY(cudaOccupancyMaxActiveBlocksPerMultiprocessor(
+        &num_blocks_init_probe_tbl, init_probe_tbl<key_type,size_type>, block_size, 0
+    ));
+
+    int dev_id {-1};
+    CUDA_TRY(cudaGetDevice(&dev_id));
+
+    int num_sms {-1};
+    CUDA_TRY(cudaDeviceGetAttribute(&num_sms, cudaDevAttrMultiProcessorCount, dev_id));
+
+    const int num_states = num_sms * std::max(num_blocks_init_build_tbl, num_blocks_init_probe_tbl) * block_size;
+    curandState *devStates;
+    CUDA_TRY(cudaMalloc(&devStates, num_states * sizeof(curandState)));
+
+    init_curand<<<(num_states - 1) / block_size + 1, block_size>>>(devStates, num_states);
+
+    CUDA_TRY(cudaGetLastError());
+    CUDA_TRY(cudaDeviceSynchronize());
+
+    key_type* build_tbl_sorted;
+    CUDA_TRY(cudaMalloc(&build_tbl_sorted, build_tbl_size * sizeof(key_type)));
+
+    size_type lottery_size = rand_max < std::numeric_limits<key_type>::max() - 1 ? rand_max + 1 : rand_max;
+    key_type* lottery;
+    bool lottery_in_device_memory = true;
+
+    size_t free_gpu_mem {0};
+    size_t total_gpu_mem {0};
+
+    CUDA_TRY(cudaMemGetInfo(&free_gpu_mem, &total_gpu_mem));
+
+    if (free_gpu_mem > lottery_size * sizeof(key_type)) {
```

Comment:
Since this is a benchmark, do you really need to make this decision? If the GPU memory is too small, run smaller benchmarks? Otherwise the GPU memory size is affecting the benchmark code.

Reply:
I agree. This check is removed in https://github.com/rapidsai/cudf/pull/2105/commits/10d4e20c9ab010c429d853125fb4cffe821554fd.